File: Languages/Lua/English/Logic/Logic.lua
-- 1. Variables and Data Types
-- Variables are like "boxes" where we store information. Each box has a specific type.

age = 25            -- Integer type: stores whole numbers
height = 1.75       -- Floating-point type: stores numbers with decimal places
name = "John"       -- String type: stores a sequence of characters
student = true      -- Boolean type: stores true or false

print("Age: " .. age)
print("Height: " .. height)
print("Name: " .. name)
print("Student: " .. tostring(student))

-- 2. Operators
-- Arithmetic, comparison, and logical operators.

-- Arithmetic Operators
sum = 10 + 5  -- Addition
remainder = 10 % 3 -- Modulo (remainder of division)
print("Sum: " .. sum)
print("Remainder: " .. remainder)

-- Comparison Operators
result = 10 > 5  -- Checks if 10 is greater than 5
equal = 5 == 5   -- Checks if 5 is equal to 5
print("10 > 5: " .. tostring(result))
print("5 == 5: " .. tostring(equal))

-- Logical Operators
andCondition = (10 > 5) and (5 < 10)  -- 'AND' logical: both conditions must be true
orCondition = (10 > 5) or (5 > 10)    -- 'OR' logical: at least one condition must be true
print("AND condition: " .. tostring(andCondition))
print("OR condition: " .. tostring(orCondition))

-- 3. Conditional Structures
-- Conditions to make decisions in the program.

personAge = 20
if personAge >= 18 then
    print("You are an adult.")
else
    print("You are a minor.")
end

-- Switch Case: an alternative to if-else when there are multiple possible conditions
number = 2
if number == 1 then
    print("It is one")
elseif number == 2 then
    print("It is two")
else
    print("Another number")
end

-- 4. Loops
-- Used when we want to repeat a block of code multiple times.

for i = 1, 5 do
    print(i)  -- Prints 1, 2, 3, 4, 5
end

j = 1
while j <= 5 do
    print(j)  -- Prints 1, 2, 3, 4, 5
    j = j + 1  -- Increments the value of j
end

k = 1
repeat
    print(k)  -- Prints 1, 2, 3, 4, 5
    k = k + 1
until k > 5

-- 5. Functions (Methods)
-- Functions are used to organize and reuse blocks of code.

function greeting()
    print("Hello, welcome!")
end

function sum(a, b)
    return a + b  -- Returns the sum of two numbers
end

greeting()  -- Calls the greeting function
sumResult = sum(5, 3)  -- Calls the sum function
print("Sum result: " .. sumResult)

-- 6. Input and Output
-- Used to interact with the user, capturing inputs and displaying outputs.

function inputOutput()
    io.write("Enter your name: ")  -- Captures the user's name
    name = io.read()
    print("Hello, " .. name)
end

-- 7. String Manipulation
-- Strings are sequences of characters, useful for working with text.

function stringManipulation()
    name = "John"
    greetingMessage = "Hello, " .. name .. "!"  -- Concatenating strings
    print(greetingMessage)  -- Displays: Hello, John!

    word = "programming"
    length = #word  -- Gets the length of the string
    contains = string.find(word, "gram") ~= nil  -- Checks if the word contains "gram"
    print("Word length: " .. length)
    print("Contains 'gram': " .. tostring(contains))
end

-- 8. Arrays
-- Arrays are used to store multiple values of the same type in a single variable.

function arrayExample()
    numbers = {1, 2, 3, 4, 5}  -- Array declaration and initialization
    print("First number: " .. numbers[1])  -- Accesses the first value
    print("Last number: " .. numbers[#numbers])  -- Accesses the last value
end

-- 9. Classes and Objects
-- Creating a class and instantiating objects from it.

Person = {}
Person.__index = Person

function Person:new(name, age)
    local person = setmetatable({}, self)
    person.name = name
    person.age = age
    return person
end

function Person:introduce()
    print("Hello, my name is " .. self.name .. " and I am " .. self.age .. " years old.")
end

function classObjectExample()
    person1 = Person:new("Nick", 30)  -- Creating an object of the Person class
    person1:introduce()  -- Calling the method
end

-- 10. Concept of `self`
-- `self` is used to refer to the current object inside a class method.

function selfExample()
    Car = {}
    Car.__index = Car

    function Car:new(model)
        local car = setmetatable({}, self)
        car.model = model
        return car
    end

    function Car:showModel()
        print("Car model: " .. self.model)
    end

    car = Car:new("Beetle")
    car:showModel()  -- Displays: Car model: Beetle
end

-- 11. Inheritance
-- Creating a child class that inherits from a parent class.

Animal = {}
Animal.__index = Animal

function Animal:new(name)
    local animal = setmetatable({}, self)
    animal.name = name
    return animal
end

function Animal:makeSound()
    print("The animal makes a sound.")
end

Dog = setmetatable({}, Animal)
Dog.__index = Dog

function Dog:makeSound()
    print("The dog barks.")
end

function inheritanceExample()
    dog = Dog:new("Rex")
    dog:makeSound()  -- Calls the overridden method
end

-- 12. Encapsulation
-- Using access modifiers to protect data.

function encapsulationExample()
    BankAccount = {}
    BankAccount.__index = BankAccount

    function BankAccount:new()
        local account = setmetatable({}, self)
        account._balance = 0  -- Private attribute
        return account
    end

    function BankAccount:deposit(amount)
        if amount > 0 then
            self._balance = self._balance + amount
        end
    end

    function BankAccount:getBalance()
        return self._balance  -- Access the balance through a public method
    end

    account = BankAccount:new()
    account:deposit(500)
    print("Account balance: " .. account:getBalance())
end

-- 13. Exceptions
-- Handling errors during program execution.

function exceptionExample()
    pcall(function()
        result = 10 / 0  -- Division by zero
    end)
end

-- 14. Working with `ArrayList`
-- The ArrayList is a collection that can dynamically grow.

function arrayListExample()
    list = {"Java", "Python", "C++"}

    print("List size: " .. #list)  -- Displays the list size
    print("First item: " .. list[1])  -- Accesses the first item of the list
end

File: Languages/Swift/English/Logic/Logic.swift
// 1. Variables and Data Types
let age = 25            // Integer type: stores whole numbers
let height = 1.75       // Float type: stores numbers with decimals
let name = "John"       // String type: stores a sequence of characters
let isStudent = true    // Boolean type: stores true or false

print("Age: \(age)")
print("Height: \(height)")
print("Name: \(name)")
print("Student: \(isStudent)")

// 2. Operators
let sum = 10 + 5  // Addition
let remainder = 10 % 3 // Modulo (remainder of division)
print("Sum: \(sum)")
print("Remainder: \(remainder)")

// Comparison Operators
let isGreaterThan = 10 > 5  // Checks if 10 is greater than 5
let areEqual = 5 == 5       // Checks if 5 is equal to 5
print("10 > 5: \(isGreaterThan)")
print("5 == 5: \(areEqual)")

// Logical Operators
let andCondition = (10 > 5) && (5 < 10)  // Logical AND: both conditions must be true
let orCondition = (10 > 5) || (5 > 10)   // Logical OR: at least one condition must be true
print("AND Condition: \(andCondition)")
print("OR Condition: \(orCondition)")

// 3. Conditional Structures
let personAge = 20
if personAge >= 18 {
    print("You are an adult.")
} else {
    print("You are a minor.")
}

// Switch Case
let number = 2
switch number {
case 1:
    print("It is one")
case 2:
    print("It is two")
default:
    print("Another number")
}

// 4. Loops (Iteration)
for i in 1...5 {
    print(i)  // Prints 1, 2, 3, 4, 5
}

var j = 1
while j <= 5 {
    print(j)  // Prints 1, 2, 3, 4, 5
    j += 1    // Increments the value of j
}

var k = 1
repeat {
    print(k)  // Prints 1, 2, 3, 4, 5
    k += 1
} while k <= 5

// 5. Functions (Methods)
func greet() {
    print("Hello, welcome!")
}

func add(a: Int, b: Int) -> Int {
    return a + b  // Returns the sum of the two numbers
}

greet()  // Calls the greet function
let sumResult = add(a: 5, b: 3)  // Calls the add function
print("Sum result: \(sumResult)")

// 6. Input and Output
func inputOutput() {
    print("Enter your name: ")
    if let name = readLine() {
        print("Hello, \(name)")
    }
}

// 7. String Manipulation
func stringManipulation() {
    let name = "John"
    let greeting = "Hello, \(name)!"  // Concatenating strings
    print(greeting)  // Prints: Hello, John!

    let word = "programming"
    let length = word.count  // Checks the length of the string
    let contains = word.contains("gram")  // Checks if the word contains "gram"
    print("Word length: \(length)")
    print("Contains 'gram': \(contains)")
}

// 8. Arrays
func arrayExample() {
    let numbers = [1, 2, 3, 4, 5]  // Declaration and initialization of an array
    print("First number: \(numbers[0])")  // Accesses the first value
    print("Last number: \(numbers[numbers.count - 1])")  // Accesses the last value
}

// 9. Classes and Objects
class Person {
    var name: String
    var age: Int

    init(name: String, age: Int) {
        self.name = name
        self.age = age
    }

    func introduce() {
        print("Hello, my name is \(self.name) and I am \(self.age) years old.")
    }
}

func classObjectExample() {
    let person1 = Person(name: "Nico", age: 30)  // Creating an object of the Person class
    person1.introduce()  // Calls the method
}

// 10. The `self` Concept
func selfExample() {
    class Car {
        var model: String

        init(model: String) {
            self.model = model  // 'self' refers to the class attribute
        }

        func showModel() {
            print("Car model: \(self.model)")
        }
    }

    let car = Car(model: "Beetle")
    car.showModel()  // Prints: Car model: Beetle
}

// 11. Inheritance
class Animal {
    var name: String

    init(name: String) {
        self.name = name
    }

    func makeSound() {
        print("The animal makes a sound.")
    }
}

class Dog: Animal {
    override func makeSound() {
        print("The dog barks.")
    }
}

func inheritanceExample() {
    let dog = Dog(name: "Rex")
    dog.makeSound()  // The overridden method is called
}

// 12. Encapsulation
func encapsulationExample() {
    class BankAccount {
        private var balance: Double  // Private attribute

        init() {
            self.balance = 0
        }

        func deposit(amount: Double) {
            if amount > 0 {
                self.balance += amount
            }
        }

        func getBalance() -> Double {
            return self.balance  // Access to balance through a public method
        }
    }

    let account = BankAccount()
    account.deposit(amount: 500)
    print("Account balance: \(account.getBalance())")
}

// 13. Exceptions
func exceptionExample() {
    do {
        let result = 10 / 0  // Division by zero
    } catch {
        print("Error: \(error)")
    }
}

// 14. Working with Arrays
func arrayListExample() {
    var list = ["Java", "Python", "C++"]

    print("List size: \(list.count)")  // Displays the size of the list
    print("First item: \(list[0])")  // Accesses the first item of the list
}
